sample/board.py

```diff
@@ -34,11 +36,59 @@ def sub_row_as_str(self, row, sub_row):
             cols.append(self.sub_boards[row][col].row_as_str(sub_row))
         return "     ".join(cols)
 
+    def is_valid(self,x,y):
+        return (not self.sub_boards[x][y].is_won and not self.sub_boards[x][y].num_moves == 9
```

Comment:
```suggestion
        return (not self.sub_boards[x][y].is_won and not self.sub_boards[x][y].is_full() and valid_coordinates(x, y)
```

Reply:
As you see above, I suggest putting and x < 3 and x >= 0 and y < 3 and y >= 0 into a valid_coordinates() method